Incorrect error reported for trait impls on DST with the last field being Sized

Given the following code: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=d9a00dc0e329f0ea73b037f3af9f26db

```rust
use std::marker::PhantomData;

trait Trait {}

struct Unsized([u8], ());
struct Foo<T: ?Sized>(PhantomData<T>);

impl<T> Trait for Foo<T> {}
impl Trait for Foo<Unsized> {}
```

The current output is:

```
error[E0119]: conflicting implementations of trait `Trait` for type `Foo<Unsized>`:
 --> src/main.rs:9:1
  |
8 | impl<T> Trait for Foo<T> {}
  | ------------------------ first implementation here
9 | impl Trait for Foo<Unsized> {}
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^ conflicting implementation for `Foo<Unsized>`

```

Ideally the output should look like:

```
error[E0277]: the size for values of type `[u8]` cannot be known at compilation time
 --> src/lib.rs:1:16
  |
5 | struct Unsized([u8], ());
  |                ^^^^ doesn't have a size known at compile-time
  |
  = help: the trait `Sized` is not implemented for `[u8]`
  = note: only the last field of a struct may have a dynamically sized type
  = help: change the field's type to have a statically known size
help: borrowed types always have a statically known size
  |
5 | struct Unsized(&[u8], ());
  |                ^
help: the `Box` type always has a statically known size and allocates its contents in the heap
  |
5 | struct Unsized(Box<[u8]>, ());
  |                ^^^^    ^

error: aborting due to previous error

```

Note: this error is particularly confusing because the trait impls \*dont\* overlap after fixing the field ordering error

Fixed since 1.78.0:
```
error[E0277]: the size for values of type `[u8]` cannot be known at compilation time
 --> src/lib.rs:5:16
  |
5 | struct Unsized([u8], ());
  |                ^^^^ doesn't have a size known at compile-time
  |
  = help: the trait `Sized` is not implemented for `[u8]`
  = note: only the last field of a struct may have a dynamically sized type
  = help: change the field's type to have a statically known size
help: borrowed types always have a statically known size
  |
5 | struct Unsized(&[u8], ());
  |                +
help: the `Box` type always has a statically known size and allocates its contents in the heap
  |
5 | struct Unsized(Box<[u8]>, ());
  |                ++++    +

error[E0119]: conflicting implementations of trait `Trait` for type `Foo<Unsized>`
 --> src/lib.rs:9:1
  |
8 | impl<T> Trait for Foo<T> {}
  | ------------------------ first implementation here
9 | impl Trait for Foo<Unsized> {}
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^ conflicting implementation for `Foo<Unsized>`

Some errors have detailed explanations: E0119, E0277.
For more information about an error, try `rustc --explain E0119`.
```

The "conflicting implementations" error still appears, but I thinks that's fine, especially because rustc suggests that you should turn the `[u8]` field into a sized one (`&[u8]` or `Box<[u8]>`). So at least the diagnostics are consistent with each other.

@rustbot label: +E-needs-test

--- a/tests/ui/unsized/unsized-non-last-field-overlapping-impls-83097.rs
+++ b/tests/ui/unsized/unsized-non-last-field-overlapping-impls-83097.rs
@@ -1,0 +1,20 @@
+//! Regression test for <https://github.com/rust-lang/rust/issues/83097>.
+//!
+//! Only the `E0119` conflicting-implementations error used to be reported here, which hid the
+//! actual mistake: the unsized field is not the last field of the struct. Both errors are now
+//! emitted.
+
+use std::marker::PhantomData;
+
+trait Trait {}
+
+struct Unsized([u8], ());
+//~^ ERROR the size for values of type `[u8]` cannot be known at compilation time
+
+struct Foo<T: ?Sized>(PhantomData<T>);
+
+impl<T> Trait for Foo<T> {}
+impl Trait for Foo<Unsized> {}
+//~^ ERROR conflicting implementations of trait `Trait` for type `Foo<Unsized>`
+
+fn main() {}

--- a/tests/ui/unsized/unsized-non-last-field-overlapping-impls-83097.stderr
+++ b/tests/ui/unsized/unsized-non-last-field-overlapping-impls-83097.stderr
@@ -1,0 +1,30 @@
+error[E0277]: the size for values of type `[u8]` cannot be known at compilation time
+  --> $DIR/unsized-non-last-field-overlapping-impls-83097.rs:11:16
+   |
+LL | struct Unsized([u8], ());
+   |                ^^^^ doesn't have a size known at compile-time
+   |
+   = help: the trait `Sized` is not implemented for `[u8]`
+   = note: only the last field of a struct may have a dynamically sized type
+   = help: change the field's type to have a statically known size
+help: borrowed types always have a statically known size
+   |
+LL | struct Unsized(&[u8], ());
+   |                +
+help: the `Box` type always has a statically known size and allocates its contents in the heap
+   |
+LL | struct Unsized(Box<[u8]>, ());
+   |                ++++    +
+
+error[E0119]: conflicting implementations of trait `Trait` for type `Foo<Unsized>`
+  --> $DIR/unsized-non-last-field-overlapping-impls-83097.rs:17:1
+   |
+LL | impl<T> Trait for Foo<T> {}
+   | ------------------------ first implementation here
+LL | impl Trait for Foo<Unsized> {}
+   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^ conflicting implementation for `Foo<Unsized>`
+
+error: aborting due to 2 previous errors
+
+Some errors have detailed explanations: E0119, E0277.
+For more information about an error, try `rustc --explain E0119`.

